### DISCUSSION TITLE
How can I implement List of Card  like view ??

### DISCUSSION THREAD

**Discussion by heejit** [cogentcore/core#1243] [General]
Is there any widget which I can use to display Card Like Info and a container which can display multiple card with scrolling.

**Comment by kkoreilly** [comment]
You can make an outer container `Frame` and then inner `Frame` widgets for the cards, with text, a form, or any other widgets inside the frame. You can make your card its own widget type that embeds `core.Frame` and defines an `Init` method that configures it appropriately.

**Reply by heejit** [reply]
I am not able to get it right is there any example or document available ?

this is my code but i have nil pointer 

```
type Card struct {
	core.Frame
	lblHeading *core.Text
	lblSubHeading *core.Text
}

func (this *Card) Init() {
	this.Frame.Init()
	this.lblHeading = core.NewText(this)
	this.lblSubHeading = core.NewText(this)
}

func NewCardWidget(parent *core.Frame, h1 string, h2 string) *Card {
	v := &Card{}
	v.Frame.Parent = parent
	v.lblHeading.SetText(h1)
	v.lblSubHeading.SetText(h2)
	return v
}

```

**Reply by kkoreilly** [reply]
There is some guidance on the [advanced updating page](https://www.cogentcore.org/core/advanced/updating), which I plan to improve soon. A better approach would look something like this:

```go
type Card struct {
	core.Frame
	Heading    string
	SubHeading string
}

func (cd *Card) Init() {
	cd.Frame.Init()
	tree.AddChild(cd, func(w *core.Text) {
		w.SetType(core.TextHeadlineSmall)
		w.Updater(func() {
			w.SetText(cd.Heading)
		})
	})
	tree.AddChild(cd, func(w *core.Text) {
		w.Updater(func() {
			w.SetText(cd.SubHeading)
		})
	})
}
```

Then, you can add a `//go:generate core generate` comment directive at the start of the package and run `go generate`, which will make a `typegen.go` file with a `NewCard` method and `Set*` functions for each of the heading properties, which you can then use and chain together just like Cogent Core widgets. (Note: for now, you may need to add a `var _ tree.Node = nil` line to your file to get the initial `go generate` to work, but you can remove that after the first generate; I am planning to fix that soon).

**Reply by heejit** [reply]
I created a package called widgets

widgets/card_widget.go

then i run go generate from there it shows core generate succeeded

But there is no file called typegen.go 

```
//go:generate core generate

package widgets

var _ tree.Node = nil

type Card struct {
	core.Frame
	Heading    string
	SubHeading string
}

func (cd *Card) Init() {
	cd.Frame.Init()
	tree.AddChild(cd, func(w *core.Text) {
		w.SetType(core.TextHeadlineSmall)
		w.Updater(func() {
			w.SetText(cd.Heading)
		})
	})
	tree.AddChild(cd, func(w *core.Text) {
		w.Updater(func() {
			w.SetText(cd.SubHeading)
		})
	})
}

```

**Reply by kkoreilly** [reply]
I tested it and it works for me; are you sure you got the imports correctly added to the file? (using `goimports`) (I don't see them in what you sent)

**Reply by heejit** [reply]
It work after importing it thanks.

**Reply by heejit** [reply]
how do I layout the both heading is it cd.Styler ??

**Reply by kkoreilly** [reply]
Yes, you can do `cd.Styler` in the `Init` function if for example you want to set `s.Direction = styles.Column`.

**Reply by heejit** [reply]
it is working fine thanks again
